### PR TITLE
Add infra to skip inapplicable modules

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
meta-schema: add infra to skip inapplicable modules

Add new `activate_by_schema_keys` property to `MetaSchema` defining
the schema keys that must be present, to avoid skipping the module.
Add this new key to our schema docs.
Add documentation to `topics/module_creation`.
```

## Additional Context

SC-1132
Original draft PR: https://github.com/canonical/cloud-init/pull/1515

## Test Steps

```sh
$ lxc launch ubuntu-daily:jammy jj
$ lxc exec jj -- grep -i 'no applicable' /var/log/cloud-init.log
2022-07-15 15:15:56,292 - modules.py[INFO]: Skipping modules 'snap' because no applicable config is provided.
```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/config/schema.py
Comment (aciba90) [inline cloudinit/config/schema.py:68]: If we do not define `skip_by_schema` as `NotRequired`, then mypy is going to raise type-exceptions for `MetaSchema` instances without the new property defined (I am thinking about downstream custom modules).

For example, if we leave this as it was but with the new property:

```py
# annotations add value for development, but don't break old versions
# pyver: 3.6 -> 3.8
# pylint: disable=E1101
if sys.version_info >= (3, 8):

    class MetaSchema(typing.TypedDict):
        name: str
        id: str
        title: str
        description: str
        distros: typing.List[str]
        examples: typing.List[str]
        frequency: str
        skip_by_schema: typing.List[str]

else:
    MetaSchema = dict
# pylint: enable=E1101
```

and we run mypy with a python version >= 3.8, then we get:

```sh
$ tox -e mypy
...
cloudinit/config/cc_byobu.py:35: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_bootcmd.py:23: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apt_pipelining.py:27: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apt_configure.py:35: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apk_configure.py:56: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_schema.py:715: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_modules.py:42: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_modules.py:75: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_write_files_deferred.py:23: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
Found 62 errors in 61 files (checked 489 source files)
ERROR: InvocationError for command /cloud-init/.tox/mypy/bin/python -m mypy cloudinit/ tests/ tools/ (exited with code 1)
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
ERROR:   mypy: commands failed

```
Similarly, if we do not include `skip_by_schema: typing.List[str]` at all, then we get error from schemas with the property.

This is not a problem for us, as we could add the key to all our meta schemas. But could hit downstream projects with custom modules.

If we do not want to make use of `type_extensions`, then we could tell downstream projects:

1. Add `"skip_by_schema": []` to their custom modules.
2. Or remove this if block and define `MetaSchema = dict` as a temporal solution, but apply 1 at some point.

More context on: https://github.com/canonical/cloud-init/pull/1589

Comment (aciba90) [reply cloudinit/config/schema.py:68]: After a team discussion, we have agreed on accepting the risk and use `type_extensions` only at type-checking time in this case.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
@aciba90 looks really good and behaves well. A couple of minor comments, one related to moving the importlib changes out of this PR if we can. 

I think there could be a possibility that we may not want to continue to do the initial "skip inapplicable" modules check if cloud-config instance-data hasn't changed. But, that is a minor recurring cost per boot during module discovery.  It may be nice at some point if we grew the ability to report module status  from `cloud-init status --long [--format json]`   that could report whether the module participated in boot, or was skipped (either due to previously running or due to inactivity/inapplicable config). This is not something for this PR, but future feature question for us to discuss long-term.

medium term, maybe for separate PR as well: the write_files vs write_files_deferred case is potentially interesting. If we have `write_files: defered: true` in config we could skip running the base `cc_write_files`. This is the only case like this were specific values of config key indicates whether or not to skip the module. But, I think it's a corner case that may not be worth investing time in.

File: cloudinit/config/schema.py
Comment (aciba90) [reply cloudinit/config/schema.py]: done

File: cloudinit/config/cc_snap.py
Comment (aciba90) [reply cloudinit/config/cc_snap.py]: I agree. This renaming does also improve developer experience, as the old name allowed double negation.

Additionally, I have renamed `_is_inapplicable` to `_is_active` to avoid double negation.

File: tests/unittests/config/test_modules.py
Comment (aciba90) [reply tests/unittests/config/test_modules.py]: done

File: cloudinit/config/modules.py
Comment (aciba90) [reply cloudinit/config/modules.py:308]: I agree. We are just mimicking what we do for [skipped unverified modules](https://github.com/canonical/cloud-init/pull/1597/files#diff-96ac54f7b1a81a1112d89cffee1171b69631165e53f53495c2532b9607b5fcb9R314).

And if in the future we want to extend it, we can do so for both cases.

File: cloudinit/config/schema.py
Comment (aciba90) [reply cloudinit/config/schema.py]: done

File: tests/unittests/config/test_schema.py
Comment (aciba90) [reply tests/unittests/config/test_schema.py]: done

File: tests/unittests/config/test_schema.py
Comment (aciba90) [reply tests/unittests/config/test_schema.py]: done

File: tests/unittests/config/test_schema.py
Comment (aciba90) [reply tests/unittests/config/test_schema.py]: done

File: cloudinit/config/schema.py
Comment (aciba90) [reply cloudinit/config/schema.py:70]: > Explicit runtime checks for the win. Thanks for this to avoid trying to import typing modules when we happen to be in runtime.

Yeah!


> One thing I don't think we can do though is drop the additional check for 3,8 without some level of checking that import TypedDict is present. As it is, python 3.6 and 3.7 are still 'supported', so doesn't that leave us with exposure to typing.TypedDict not existing yet?

Maybe it went unnoticed, but we are importing `TypeDict` from `typing_extensions` which provides a back-ported implementation of `TypeDict` for 3.6 and 3.7. Therefore, we do not need additional handling.

File: tests/unittests/config/test_schema.py
Comment (aciba90) [reply tests/unittests/config/test_schema.py]: done

**Comment by aciba90** [comment]
> A couple of minor comments, one related to moving the importlib changes out of this PR if we can.

Extracted to https://github.com/canonical/cloud-init/pull/1605

> I think there could be a possibility that we may not want to continue to do the initial "skip inapplicable" modules check if cloud-config instance-data hasn't changed. But, that is a minor recurring cost per boot during module discovery. It may be nice at some point if we grew the ability to report module status from `cloud-init status --long [--format json]` that could report whether the module participated in boot, or was skipped (either due to previously running or due to inactivity/inapplicable config). This is not something for this PR, but future feature question for us to discuss long-term.

That would be interesting. Should we add it to SC-1097 or somewhere else? To not forget about its discussion.

> medium term, maybe for separate PR as well: the write_files vs write_files_deferred case is potentially interesting. If we have `write_files: defered: true` in config we could skip running the base `cc_write_files`. This is the only case like this were specific values of config key indicates whether or not to skip the module. But, I think it's a corner case that may not be worth investing time in.

If this is the only corner case, I would say that it is not worth to implement a more complex way to express logic within `activate_by_schema_keys`. I am thinking about something like:

```py
{
    "activate_by_schema_keys": ["write_files.defered == True"]
}
```

But if this necessity grows, for sure we could do it.

Thanks, @blackboxsw, for the review and constructive feedback!
I think I have covered all the open topics. Please, have a second review.

File: cloudinit/config/schema.py
Comment (blackboxsw) [inline cloudinit/config/schema.py:917]: Let's mark up the key names with ``<key_name>`` to align with other doc config key references. It makes readthedocs decorate it in red.

```suggestion
    schema_keys = ", ".join(meta["activate_by_schema_keys"])
    schema_keys = ", ".join(
        f"``{k}``" for k in meta["activate_by_schema_keys"]
    )
    return f"**Activate only on keys:** {schema_keys}\n\n"
```

**Review by blackboxsw — APPROVED** [review]
Looks really good. one minor rendered text change request (which I realize affects your unit tests). Otherwise +1 from me. I'm testing this now

File: cloudinit/config/cc_snap.py
Comment (blackboxsw) [reply cloudinit/config/cc_snap.py]: looks good.

**Comment by holmanb** [comment]
I think the commit message needs to be updated, but otherwise looks good to me!

**Review by holmanb — APPROVED** [review]

**Comment by blackboxsw** [comment]
Push the minor doc change and will land when CI passes

**Comment by aciba90** [comment]
Thanks, @blackboxsw and @holmanb, for handling and landing this.

**Pull request merged**
